client/network/sync/src/engine.rs

```diff
@@ -590,6 +608,33 @@ where
 			self.tick_timeout.reset(TICK_TIMEOUT);
 		}
 
+		// go over all connected peers and check if any of them have been idle for a while. Idle in
+		// this case means that we haven't sent or received block announcements to/from this peer.
+		// If that is the case, because of #5685, it could be that the block announces substream is
+		// not actually open and and this peer is just wasting a slot and is should be replaced with
+		// some other node that is willing to send us block announcements.
+		for (id, peer) in self.peers.iter() {
+			// because of a delay between disconnecting a peer in `SyncingEngine` and getting the
+			// response back from `Protocol`, a peer might be reported and disconnect multiple
+			// times. To prevent this from happening (until the underlying issue is fixed), keep
+			// track of evicted peers and report and disconnect them only once.
+			if self.evicted.contains(id) {
+				continue
+			}
+
+			let last_received_late =
+				peer.last_notification_received.elapsed() > INACTIVITY_EVICT_THRESHOLD;
+			let last_sent_late = peer.last_notification_sent.elapsed() > INACTIVITY_EVICT_THRESHOLD;
```

Comment:
So this has an effect of disconnecting peers if they don't learn any new blocks from us? Sounds reasonable, but is likely redundant if the only way to learn that a peer knows some block is receiving a block announcement from it.

Reply:
It's a bit hackish but I'm trying to incorporate your change from while back where we detect closed outbound substream into this change.

Consider the following. We're connected to two other peers, peerA and peerB. We receive a block announcement from peerA and since peerB has not sent this block announcement to us yet, we send it to them. This means that we won't be receiving the same block announcement from peerB but if we were able to send the block announcement to them without the connection getting closed, it means that at least the outbound substream is working and there is potential of receiving a block announcement from them some time in the future. This is not perfect because the outbound substream is independent of the inbound substream so being able to send something doesn't mean the inbound substream is not closed but if the outbound is closed, the connection has been rejected and now we're able to detect that. Us being able to send something to someone means that we're able to receive block announcements from one of the other peers.

I want to emphasize that this entire PR is a temporary attempt to unblock a stalled chain. The proper way to fix this is to not accept the substream until it has been validated by the protocol but that change is still WIP.

Reply:
I probably don't understand something. `last_sent_late` can only be true if we have not been sending block announcements to the peer for some time (note that we update `last_notification_sent` even before we sent the BA, and this doesn't detect a closed substream). For this to happen, we must be sure that the peer already knows about these blocks. The only way to know this is to receive block announcements from this peer. What makes `last_received_late` to be `false` (i.e., we heard from the peer about these blocks).

How does this work then?

Reply:
If we send a block announcement to peer, they won't send the same announcement back to us (unless they send it at the same time). Meaning if within a 2-minute-period we'd always hear about block announcements before them, they would get disconnected. Them not sending anything inside this time window, provided that our block announcements go through, doesn't mean there is no potential for receiving a block announcement in the future. Right now we can't detect if the inbound substream is closed so it's far from perfect solution. 

If you don't like this, do you have a suggestion on how to improve it? I can remove this check entirely as well if people don't like it.

Reply:
Yes `last_received_late && last_sent_late` would be triggered when we're stalled but that's more of a global state and should not be stored in `Peer` necessarily. You have a point and this felt hackish from the beginning and I'm liking it less and less.

I'll think about the implementation some more. The problem is that we can't expect to receive a block announcement from peer within `INACTIVITY_EVICT_THRESHOLD` if we've sent them a block inside that time window so the eviction may not be warranted.

Reply:
I have been playing around with different approaches but I think I will keep how it currently works unless there is more opposition.

When the node is major syncing, it's not announcing blocks so it might get incorrectly evicted even though the substream is open and there is future potential for receiving block announcements. This also partly addresses the bug in block request handler which allows the node to downloads blocks from peers who rejected. Now it will evict that peer since it won't receive any block announcements from the peer and instead looks for some other free peer who is able to provide block announcements. Block announce substreams are unidirectional between two nodes where one is major syncing and one is not so the other peer is tracking `last_received_late` and the other `last_sent_late` and since first is the consumer and the second is the producer, the substream should remain open. If the produce is only connected to nodes who major sync from it, it will evict those peers since it's own chain will otherwise stall.

The downside of this whole PR is that peers who are both major syncing will evict each other since neither sends block announcements and it's hard to detect if the remote node is major syncing. They will still remain connected to each other for 30 seconds and during that time will download blocks. If this sounds too risky, we can also disable to eviction while the node is major syncing, allowing it to stay connected for longer periods of time.

@dmitry-markin @bkchr any comments?

Reply:
This all makes sense and I am in favor of merging this PR. As for disconnecting two major syncing nodes announcing blocks to each other, this seems unlikely to break anything under normal circumstances when most of the nodes are keep-up syncing and we can reconnect to one of those. But I don't know if it can cause harm during some rare occasions — e.g., when we try to bootstrap a new network or deal with network stalling issues. May be we should just ask to reproduce the conditions of adding like 90-99% of major syncing nodes to the Versi testnet and check if and how fasts the network recovers.

Reply:
Yes I agree it won't likely be a major problem but I'll check what happens if there are multiple nodes trying to major sync from each other in a local testnet

Reply:
One thing that could also happen that one node isn't sending a block announcement, because we when we received a notification from one block, we don't send the node the same block announcement again. But maybe that is really a small corner case.

In general the idea of not evicting based on we sent them a block announcement when we are in major sync sounds like a good idea.

Reply:
I think we can merge this. I looked at the Versi testnet files and according to them, there were many of them and they all major synced successfully with the eviction patch applied so I believe it's safe to merge this PR.

Reply:
Okay. :)

Reply:
One thing I thought about, that we probably did not test is to have all the nodes running with this patch and then starting a node to sync from the other nodes? The nodes providing the blocks will disconnect the syncing node? If there are not enough nodes in the network, the syncing node could get stuck.

Reply:
Nodes that are at the tip won't disconnect the node that is major syncing because the change keeps track of last send and last received separately. The only issue arises if when two nodes are major syncing from each other and since neither of them are publishing blocks, they will evict each other, reconnect again later, and then evict again until they reach tip.

In theory it sounds a little risky which is why I was wondering if the eviction should be disabled while the node is major syncing and the enable when the tip is reached. I don't believe it to be a huge problem is practice but it's a small change so maybe worth adding anyway?

Reply:
Ty you! I also checked the changes again and what you are saying makes sense! Before I had checked, I thought we would check if receiving or sending is late, but we do receiving and sending. But now I don't get how this can disconnect any node as long as we are connected to any node that sends us announcements. Point being that we announce blocks to every body after having imported it, but this then also resets the send timer. So, this should help when we are connected only to inactive peers.

But yeah, let's merge this!

Reply:
It helps us to unstall our chain if we're connected to only inactive peers which is the common pattern in these reports we've been getting. We can incorporate another change into `NotificationService` where the protocol is informed when the inbound substream is closed and it can then decide what happens. For the syncing protocol, the peer would be evicted because it's useless to us.

I need to research what happens with other protocols because I don't see what purpose it serves to keep a node when they've closed their outbound substream to us but it's there for a reason because it was implemented.

Reply:
Yeah, as I said, let's merge this. It should not make anything worse :)